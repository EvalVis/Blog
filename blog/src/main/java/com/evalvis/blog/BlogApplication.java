package com.evalvis.blog;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.annotation.Bean;
import org.springframework.http.converter.protobuf.ProtobufHttpMessageConverter;

@SpringBootApplication
public class BlogApplication {

	@Bean
	ProtobufHttpMessageConverter protobufHttpMessageConverter() {
		return new ProtobufHttpMessageConverter();
	}

	public static void main(String[] args) {
		SpringApplication.run(BlogApplication.class, args);
	}

}
