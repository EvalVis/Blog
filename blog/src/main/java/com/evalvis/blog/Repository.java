package com.evalvis.blog;

public interface Repository<T> {
    T save(T entity);

    interface Entry {
        String getId();
    }
}
