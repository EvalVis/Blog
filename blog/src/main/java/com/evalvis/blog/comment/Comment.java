package com.evalvis.blog.comment;

public final class Comment {

    private final String author;
    private final String content;
    private final String postId;

    public Comment(String author, String content, String postId) {
        this.author = author;
        this.content = content;
        this.postId = postId;
    }

    public CommentRepository.CommentEntry save(CommentRepository commentRepository) {
        return commentRepository.save(new CommentRepository.CommentEntry(author, content, postId));
    }

    public String getAuthor() {
        return author;
    }

    public String getContent() {
        return content;
    }

    public String getPostId() {
        return postId;
    }
}
