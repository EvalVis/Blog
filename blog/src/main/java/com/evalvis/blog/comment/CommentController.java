package com.evalvis.blog.comment;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("comments")
final class CommentController {

    private final CommentRepository commentRepository;

    @Autowired
    CommentController(@Qualifier("mongoDbCommentRepository") CommentRepository commentRepository) {
        this.commentRepository = commentRepository;
    }

    @PostMapping(value = "/create")
    ResponseEntity<CommentRepository.CommentEntry> create(@RequestBody Comment comment)
    {
        return ResponseEntity.ok(comment.save(commentRepository));
    }

    @GetMapping(value = "/list-comments/{postId}")
    ResponseEntity<List<CommentRepository.CommentEntry>> listCommentsOfPost(
            @PathVariable String postId
    ) {
        return ResponseEntity.ok(commentRepository.findCommentEntriesByPostEntryId(postId));
    }
}
