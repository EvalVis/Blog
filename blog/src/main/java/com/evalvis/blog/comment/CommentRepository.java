package com.evalvis.blog.comment;

import com.evalvis.blog.Repository;
import com.fasterxml.jackson.annotation.JsonPropertyOrder;

import java.util.List;

public interface CommentRepository extends Repository<CommentRepository.CommentEntry> {

    List<CommentRepository.CommentEntry> findCommentEntriesByPostEntryId(String postId);

    interface CommentEntry extends Entry {
        String getAuthor();
        String getContent();
        String getPostEntryId();
    }
}
