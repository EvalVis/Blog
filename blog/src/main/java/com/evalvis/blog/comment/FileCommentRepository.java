package com.evalvis.blog.comment;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonPropertyOrder;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.stereotype.Repository;

import java.io.*;
import java.nio.file.Paths;
import java.util.List;
import java.util.UUID;
import java.util.function.Supplier;
import java.util.stream.Collectors;

@Repository("fileCommentRepository")
public class FileCommentRepository implements CommentRepository {

    private final Supplier<File> commentRepo;

    public FileCommentRepository() {
        commentRepo = FileCommentRepository::initRepo;
    }

    private static File initRepo() {
        File repo = Paths.get("repo/commentRepository.json").toFile();
        if(!repo.exists()) {
            try {
                repo.createNewFile();
                try(BufferedWriter writer = new BufferedWriter(new FileWriter(repo))) {
                    writer.write("[]");
                }
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }
        return repo;
    }

    @Override
    public List<CommentRepository.CommentEntry> findCommentEntriesByPostEntryId(String postId) {
        try {
            List<CommentEntry> comments = new ObjectMapper()
                    .readValue(commentRepo.get(), new TypeReference<>() {});
            return comments
                    .stream()
                    .filter(comment -> comment.getPostEntryId().equals(postId))
                    .collect(Collectors.toList());
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public CommentRepository.CommentEntry save(CommentRepository.CommentEntry entry) {
        try {
            List<CommentEntry> comments = new ObjectMapper()
                    .readValue(commentRepo.get(), new TypeReference<>() {});
            comments.add((CommentEntry) entry);
            new ObjectMapper().writeValue(commentRepo.get(), comments);
            return entry;
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    @JsonPropertyOrder({"author", "content", "id", "postEntryId"})
    public static final class CommentEntry implements CommentRepository.CommentEntry {

        private final String id;
        private final String author;
        private final String content;
        private final String postEntryId;

        @JsonCreator
        CommentEntry(
                @JsonProperty("author") String author, @JsonProperty("content") String content,
                @JsonProperty("postEntryId") String postEntryId
        ) {
            id = UUID.randomUUID().toString();
            this.author = author;
            this.content = content;
            this.postEntryId = postEntryId;
        }

        @Override
        public String getId() {
            return id;
        }

        @Override
        public String getAuthor() {
            return author;
        }

        @Override
        public String getContent() {
            return content;
        }
        @Override
        public String getPostEntryId() {
            return postEntryId;
        }
    }
}
