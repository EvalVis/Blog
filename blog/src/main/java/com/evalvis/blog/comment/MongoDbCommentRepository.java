package com.evalvis.blog.comment;

import com.fasterxml.jackson.annotation.JsonPropertyOrder;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.repository.MongoRepository;
import org.springframework.stereotype.Repository;

@Repository("mongoDbCommentRepository")
public interface MongoDbCommentRepository extends CommentRepository,
        MongoRepository<MongoDbCommentRepository.CommentEntry, String> {

    @Document("comment")
    @JsonPropertyOrder(alphabetic=true)
    class CommentEntry implements CommentRepository.CommentEntry {

        @Id
        @Indexed(unique = true)
        private String id;
        private String author;
        private String content;
        private String postEntryId;

        public CommentEntry(String author, String content, String postEntryId) {
            this.author = author;
            this.content = content;
            this.postEntryId = postEntryId;
        }

        public CommentEntry() {
        }

        @Override
        public String getId() {
            return id;
        }

        @Override
        public String getAuthor() {
            return author;
        }

        @Override
        public String getContent() {
            return content;
        }

        @Override
        public String getPostEntryId() {
            return postEntryId;
        }
    }
}
