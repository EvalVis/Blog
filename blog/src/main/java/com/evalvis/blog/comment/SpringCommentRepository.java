package com.evalvis.blog.comment;

import com.fasterxml.jackson.annotation.JsonPropertyOrder;
import jakarta.persistence.*;
import org.springframework.context.annotation.Primary;
import org.springframework.data.repository.CrudRepository;
import org.springframework.stereotype.Repository;

@Repository("springCommentRepository")
@Primary
public interface SpringCommentRepository extends CommentRepository,
        CrudRepository<SpringCommentRepository.CommentEntry, String> {

    @Entity(name = "comment")
    @JsonPropertyOrder(alphabetic=true)
    class CommentEntry implements CommentRepository.CommentEntry {

        @Id
        @GeneratedValue(strategy = GenerationType.UUID)
        @Column(unique = true)
        private String id;
        @Column(nullable = false)
        private String author;
        @Column(nullable = false)
        private String content;
        @Column(nullable = false)
        private String postEntryId;

        CommentEntry(String author, String content, String postEntryId) {
            this.author = author;
            this.content = content;
            this.postEntryId = postEntryId;
        }

        public CommentEntry() {}

        @Override
        public String getId() {
            return id;
        }

        @Override
        public String getAuthor() {
            return author;
        }

        @Override
        public String getContent() {
            return content;
        }

        @Override
        public String getPostEntryId() {
            return postEntryId;
        }
    }
}
