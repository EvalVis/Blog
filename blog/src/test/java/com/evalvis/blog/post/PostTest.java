package com.evalvis.blog.post;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import au.com.origin.snapshots.junit5.SnapshotExtension;

import com.evalvis.blog.Repository;
import com.evalvis.blog.comment.CommentRepository;
import com.evalvis.blog.comment.FileCommentRepository;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.web.client.TestRestTemplate;

import protobufs.CommentRequest;
import protobufs.PostRequest;

import java.util.Arrays;
import java.util.stream.Collectors;

import static shadow.org.assertj.core.api.AssertionsForClassTypes.assertThat;

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@ExtendWith({SnapshotExtension.class})
public class PostTest {

    private Expect expect;

    @Value(value="${local.server.port}")
    private int port;

    @Autowired
    private TestRestTemplate restTemplate;

    @Test
    @SnapshotName("createsPostWithComments")
    public void listsPostComments() throws JsonProcessingException {
        PostRequest postRequest = PostRequest
                .newBuilder()
                .setAuthor("Human")
                .setTitle("Testing matters")
                .setContent("You either test first, test along coding, or don't test at all.")
                .build();
        PostRepository.PostEntry postFromResponse = restTemplate.postForObject(
                "http://localhost:" + port + "/posts/create",
                postRequest, SpringPostRepository.PostEntry.class
        );
        int commentCount = 2;
        String[] commentIds = new String[commentCount];

        for(int i = 0; i < commentCount; i++) {
            CommentRepository.CommentEntry createdComment = restTemplate.postForObject(
                    "http://localhost:" + port + "/comments/create",
                    CommentRequest
                            .newBuilder()
                            .setAuthor("author" + i)
                            .setContent("content" + i)
                            .setPostId(postFromResponse.getId())
                            .build(),
                    FileCommentRepository.CommentEntry.class
            );
            commentIds[i] = createdComment.getId();
        }
        CommentRepository.CommentEntry[] commentsByPost = restTemplate.getForObject(
                "http://localhost:" + port + "/comments/list-comments/" + postFromResponse.getId(),
                FileCommentRepository.CommentEntry[].class
        );

        assertThat(commentsByPost.length).isEqualTo(2);
        assertThat(
                Arrays.stream(commentsByPost).map(Repository.Entry::getId).toArray()
        ).isEqualTo(commentIds);
        expect.toMatchSnapshot(
                jsonWithMaskedProperties(commentsByPost, "id", "postEntryId")
        );
    }

    private <T> String jsonWithMaskedProperties(
            T[] objects, String... properties
    ) throws JsonProcessingException {
        ArrayNode node = new ObjectMapper().valueToTree(objects);
        node.forEach(element ->
            Arrays
                    .stream(properties)
                    .forEach(property -> ((ObjectNode) element).put(property, "#hidden#"))
        );
        return new ObjectMapper().writeValueAsString(node);
    }
}
