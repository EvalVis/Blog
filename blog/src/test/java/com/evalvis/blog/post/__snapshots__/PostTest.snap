createsPostWithComments=[
[{"author":"author0","content":"content0","postEntryId":"#hidden#","id":"#hidden#"},{"author":"author1","content":"content1","postEntryId":"#hidden#","id":"#hidden#"}]
]